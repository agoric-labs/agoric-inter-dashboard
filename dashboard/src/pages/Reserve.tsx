import { PageHeader } from '@/components/PageHeader';
import { PageContent } from '@/components/PageContent';
import { ReserveSummary } from '@/components/ReserveSummary';
import { ReserveHistory } from '@/components/ReserveHistory';

export const Reserve = () => (
  <>
    <PageHeader title="Reserve Assets" />
    <PageContent>
      <ReserveSummary />
      <ReserveHistory />
    </PageContent>
  </>
);
