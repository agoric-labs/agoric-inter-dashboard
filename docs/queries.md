## Check integrity

```sql
with mm as (
   select min(block_height) mn, max(block_height) mx
     from `xxx.raw_cosmos_hub.blocks`
)
select avg(events.count / b.event_count) as status_events
     , avg(msgs.count / b.message_count) as status_msgs
     , avg(txs.count / b.transaction_count) as status_txs
     , count(block_height) / count(distinct block_height) as status_blocks
     , count(block_height) / (max(block_height) - min(block_height)) as status_missing_total
     , sum(events.count) event_count
     , sum(msgs.count) msg_count
     , sum(txs.count) tx_count
     , count(block_height) as total_count
     , (max(block_height) - min(block_height)) as expected_total
     , (max(block_time) - min(block_time)) / count(block_height) avg_build_block_time
     , (max(_sdc_batched_at) - min(_sdc_batched_at)) / count(block_height) avg_extract_block_time
     , max(block_height) as last_block_height
     , max(block_height - prev_block_height) as max_hole
from (
  select block_height
       , block_time
       , event_count
       , message_count
       , transaction_count
       , _sdc_batched_at
       , lag(block_height) over (order by block_height) as prev_block_height
    from `xxx.raw_cosmos_hub.blocks`
    where block_height between (select mn from mm) + 500 and (select mx from mm) - 500
) b
left join (
  select block_height, count(block_height) count
    from `xxx.raw_cosmos_hub.events`
   group by 1
) events using (block_height)
left join (
  select block_height, count(block_height) count
    from `xxx.raw_cosmos_hub.messages`
   group by 1
) msgs using (block_height)
left join (
  select block_height, count(block_height) count
    from `xxx.raw_cosmos_hub.transactions`
   group by 1
) txs using (block_height)
```

## Fill missing blocks

```
kubectl port-forward agoric-mainnet-tendermint-head-5f876fb589-cvt5c 3333:3333
```

```
with mm as (
   select min(block_height) mn, max(block_height) mx
     from `xxx.raw_cosmos_hub.blocks`
)
select array_to_string(array_agg('curl -d \'' || to_json_string(to_json(struct(earliest, latest))) || '\' http://localhost:3333'), '\n'), count(earliest)
  from (
    select next as earliest
         , least(next + 99, block_height - 1) as latest
     from (
      select block_height
           , lag(block_height) over (order by block_height) as prev_block_height
        from `xxx.raw_cosmos_hub.blocks`
       where block_height between (select mn from mm) + 500 and (select mx from mm) - 500
     )
     cross join unnest(generate_array(prev_block_height + 1, block_height - 1, 100)) next
     where block_height - prev_block_height > 1
     order by 1
  )
```
