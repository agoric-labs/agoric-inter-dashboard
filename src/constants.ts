export const GET_MODULE_ACCOUNTS_URL = 'https://main-a.api.agoric.net:443/cosmos/auth/v1beta1/module_accounts';

export const GET_ACCOUNT_BALANCE_URL = (address: string) => `https://main-a.api.agoric.net:443/cosmos/bank/v1beta1/balances/${address}`;

export const GET_INTERCHAIN_BALANCES_URL = 'https://rickhouse.agoric-core.workers.dev/inter-balances';

export const UIST_DENOMINATION = 'uist';

export const VBANK_RESERVE_ACCOUNT = 'vbank/reserve';

export const GRAPH_DAYS = 90;

export const SUBQUERY_URL = 'https://index-api.onfinality.io/sq/agoric-labs/agoric-mainnet-v2';

export const enum VAULT_STATES {
  ACTIVE = 'active',
  LIQUIDATED = 'liquidated',
  LIQUIDATING = 'liquidating',
  CLOSED = 'closed',
}
