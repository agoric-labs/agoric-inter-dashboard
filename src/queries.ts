import { getDateKey, range } from "./utils";

export const PSM_DASHBOARD_QUERY = `
query {
    _metadata {
        lastProcessedHeight
    }
    psmMetrics {
        nodes {
            blockHeight
            denom
            mintedPoolBalance
        }
    }
    psmGovernances {
        nodes {
            mintLimit
            denom
        }
    }
    boardAuxes {
        nodes {
            allegedName
            decimalPlaces
        }
    }
}`;

export const PSM_GRAPH_TOKENS_QUERY = `
query {
    psmMetrics {
        nodes {
            id
            denom
        }
    }
}`;

export const PSM_TOKEN_DAILY_MINT_QUERY = (tokens: Array<string>) => `
query {
    ${tokens?.map((token: string) => `
        ${token}: psmMetricsDailies(first: 90, filter: {denom: {equalTo: "${token}"}}, orderBy:DATE_KEY_DESC) {
        nodes {
                id
                denom
                dateKey
                blockTimeLast
                totalMintedProvidedLast
            }
        }`
)}
}`
const currentDate = new Date();
export const VAULTS_DASHBOARD_QUERY = `
query {
    vaultManagerMetrics {
        nodes {
            id
            numActiveVaults
            totalCollateral
            totalDebt
            liquidatingCollateralBrand
        }
    }
    oraclePrices {
        nodes {
            priceFeedName
            typeOutAmount
            typeInAmount
        }
    }
    boardAuxes {
        nodes {
            allegedName
            decimalPlaces
        }
    }
    vaultManagerGovernances {
        nodes {
            id
            debtLimit
            liquidationMarginNumerator
            liquidationMarginDenominator
        }
    }
    vaults(filter: {state: {equalTo: "active"}}) {
        nodes {
            id
            denom
            balance
            state
            debt
            blockTime
        }
        totalCount
    }
    oraclePriceDailies(filter: {
        dateKey: { 
            in: [${getDateKey(currentDate).key}, ${getDateKey(currentDate, 1).key}] 
        }
    }) {
        nodes {
            typeInName
            typeInAmountLast
            typeOutAmountLast
            dateKey
        }
    }
}`;

export const OPEN_VAULTS_NEXT_PAGES_QUERY = (pages: number) => `
query {
    ${range(pages).map(index => `v${index}:vaults(filter: {state: {equalTo: "active"}}, first: 100, offset:${(index + 1) * 100}) {
        nodes {
            id
            denom
            balance
            state
            debt
            blockTime
        }
    }`)}
}`

export const VAULTS_DAILY_METRICS_QUERY = (tokens: string[]) => `
query {
    ${tokens?.map((token) =>
    `${token}: vaultManagerMetricsDailies(first: 90, filter:{liquidatingCollateralBrand: {equalTo: "${token}"} }, orderBy:DATE_KEY_DESC)  {
        nodes {
            id
            dateKey
            liquidatingCollateralBrand
            blockTimeLast
            totalCollateralLast
            metricsCount
            totalDebtLast
        }
    }
    ${token}_oracle: oraclePriceDailies (first: 90, filter:{typeInName: {equalTo: "${token}"}}, orderBy:DATE_KEY_DESC ) {
        nodes {
            id
            dateKey
            blockTimeLast
            typeInName
            typeInAmountLast
            typeOutAmountLast
        }
    }`)}
}`

export const RESERVE_DASHBOARD_QUERY = `
query {
    reserveMetrics {
        nodes {
            shortfallBalance
            allocations {
                nodes {
                    id
                    denom
                    value
                }
            }
        }
    }
    oraclePrices {
        nodes {
            typeInName
            typeOutAmount
            typeInAmount
        }
    }
    boardAuxes {
        nodes {
            allegedName
            decimalPlaces
        }
    }
}`


export const RESERVE_GRAPH_TOKENS_QUERY = `
query {
    reserveMetrics {
        nodes {
            allocations {
                nodes {
                    id
                    denom
                }
            }
        }
    }
}`

export const RESERVE_DAILY_METRICS_QUERY = (tokens: string[], startDate: number) => `
query {
    ${tokens?.map((token) =>
    `${token}: reserveAllocationMetricsDailies (first: 90, filter: {denom: {equalTo: "${token}"}}, orderBy:DATE_KEY_DESC) {
        nodes {
            id
            blockTimeLast
            dateKey
            valueLast
            denom
        }
    }
    ${token}_oracle: oraclePriceDailies (first: 90, filter: {typeInName: {equalTo: "${token}"}}, orderBy:DATE_KEY_DESC) {
        nodes {
          dateKey
          blockTimeLast
          typeInName
          typeInAmountLast
          typeOutAmountLast
        }
    }
    ${token}_last: reserveAllocationMetricsDailies (first: 1, filter: {denom: {equalTo: "${token}"}, valueLast: {greaterThan: "0" }, dateKey: {lessThan: ${startDate}}}, orderBy:DATE_KEY_DESC) {
        nodes {
          id
          blockTimeLast
          dateKey
          valueLast
          denom
        }
    }`)}
}`

export const INTER_DASHBOARD_QUERY = `
query {
    psmMetrics {
        nodes {
            denom
            mintedPoolBalance
            anchorPoolBalance
            mintedPoolBalance
        }
    }
    psmGovernances {
        nodes {
            denom
            mintLimit
        }
    }
    vaultManagerMetrics {
        nodes {
            id
            liquidatingCollateralBrand
            liquidatingDebtBrand
            totalDebt
            totalCollateral
        }
    }
    vaultManagerGovernances {
        nodes {
            id
            debtLimit
        }
    }
    reserveMetrics {
        nodes {
            totalFeeBurned
            shortfallBalance
            allocations {
                nodes {
                    id
                    denom
                    value
                }
            }
        }
    }
    wallets {
        totalCount
    }
    oraclePrices {
        nodes {
            id
            typeInName
            typeOutName
            typeInAmount
            typeOutAmount
            priceFeedName
        }
    }
    boardAuxes {
        nodes {
            allegedName
            decimalPlaces
        }
    }
}`

export const LIQUIDATIONS_DASHBOARD = `
query {
    boardAuxes {
        nodes {
            allegedName
            decimalPlaces
        }
    }
    vaultManagerMetrics {
        nodes {
            id
            numLiquidationsCompleted
            liquidatingCollateralValue
            numActiveVaults
            numLiquidationsCompleted
            numLiquidationsAborted
            numLiquidatingVaults
        }
    }
    vaultLiquidations (filter: {state: {equalTo: "liquidated"}}) {
        nodes {
            id
            denom
            debt
            state
            balance
            blockTime
            oraclePrice
            vaultManagerGovernance
            currentState {
                id
                denom
                debt
                state
                balance
                blockTime
            }
            liquidatingState {
                id
                denom
                debt
                state
                balance
                blockTime
            }
        }
    }
}`;

export const VAULT_STATE_DAILIES_QUERY = `
query {
    vaultStatesDailies  ( first: 90, orderBy: ID_DESC, offset: 1) {
        nodes {
            id
            blockTimeLast
            active
            closed
            liquidated
            liquidating
            liquidatedClosed
        }
    }
}`
