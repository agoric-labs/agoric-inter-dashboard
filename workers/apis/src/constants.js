export const balancesKey = 'balances';
export const BASE_NODE_URL = 'https://main-a.api.agoric.net:443';
export const SUBQUERY_URL = 'https://index-api.onfinality.io/sq/agoric-labs/agoric-mainnet-v2';
