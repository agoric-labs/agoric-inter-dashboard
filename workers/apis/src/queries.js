import { range } from './utils';

export const graphqlQuery = {
  query: `
    {
      _metadata{
        lastProcessedHeight
        lastProcessedTimestamp
      }
      oraclePrices {
        nodes {
          typeInName
          typeOutName
          typeInAmount
          typeOutAmount
          blockTime
          blockHeight
        }
      }
      oraclePriceDailies (filter: {dateKey: { greaterThan: 20240501 } }, orderBy: DATE_KEY_DESC) {
        nodes {
          typeInName
          typeOutName
          typeInAmountLast
          typeOutAmountLast
          blockTimeLast
          blockHeightLast
        }
        totalCount
      }
      vaultManagerMetrics {
        nodes {
          id
          blockTime
          blockHeight
          totalCollateral
          totalDebt
          totalCollateralSold
          totalShortfallReceived
          liquidatingCollateralValue
          liquidatingDebtBrand
          liquidatingCollateralBrand
        }
      }
      vaultManagerGovernances  {
        nodes {
          id
          blockTime
          blockHeight
          debtLimit
          liquidationMarginNumerator
          liquidationMarginDenominator
        }
      }
      vaults (filter: {state: {equalTo: "active"}}) {
        nodes {
          id
          blockTime
          blockHeight
          denom
          balance
          state
          debt
          lockedValue
          coin
        }
        totalCount
      }
      vaultLiquidations (filter: {state: {equalTo: "liquidated"}}) {
        nodes {
            id
            blockTime
            blockHeight
            denom
            debt
            state
            balance
            currentState {
                id
                denom
                debt
                state
                balance
                blockTime
            }
            liquidatingState {
                id
                denom
                debt
                state
                balance
                blockTime
            }
        }
        totalCount
    }
    }`,
};

export const PAGINATINATED_DATA_QUERIES = (vaultsPagesCount, vaultLiquidationsPagesCount, oraclePriceDailiesPagesCount) => ({
  query: `{
    ${range(vaultsPagesCount).map(
      (index) => `vaults_${index}:vaults (filter: {state: {equalTo: "active"}}, first: 100, offset:${(index + 1) * 100}) {
      nodes {
        id
        blockTime
        blockHeight
        denom
        balance
        state
        debt
        lockedValue
        coin
      }
    }`
    )}
    ${range(vaultLiquidationsPagesCount).map(
      (index) => `vaultLiquidations_${index}:vaultLiquidations (filter: {state: {equalTo: "liquidated"}}, first: 100, offset:${
        (index + 1) * 100
      }) {
      nodes {
        id
        blockTime
        blockHeight
        denom
        debt
        state
        balance
        currentState {
            id
            denom
            debt
            state
            balance
            blockTime
        }
        liquidatingState {
            id
            denom
            debt
            state
            balance
            blockTime
        }
      }
    }`
    )}    
    ${range(oraclePriceDailiesPagesCount).map(
      (
        index
      ) => `oraclePriceDailies_${index}:oraclePriceDailies (filter:  {dateKey: { greaterThan: 20240501 } }, orderBy: DATE_KEY_DESC, first: 100, offset:${
        (index + 1) * 100
      }) {
      nodes {
        typeInName
        typeOutName
        typeInAmountLast
        typeOutAmountLast
        blockTimeLast
        blockHeightLast
      }
    }`
    )}
  }`,
});

export const LIQUIDATION_ORACLE_PRICES_DAILIES_QUERY = (tokens) => ({
  query: `
{
    ${Object.entries(tokens).map(
      ([token, oracleKeys]) =>
        `${token}:oraclePriceDailies (filter: {
        dateKey: { in: [${oracleKeys.join(', ')}] }, typeInName: {equalTo: "${token}"}
    }) {
        nodes {
            id
            typeInName
            typeOutName
            typeInAmountLast
            typeOutAmountLast
            dateKey
        }
    }`
    )}
}`,
});
