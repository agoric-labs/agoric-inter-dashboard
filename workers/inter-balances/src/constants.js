export const balancesKey = 'balances';
export const BASE_NODE_URL = 'https://main-a.api.agoric.net:443';
export const SUBQUERY_URL = 'https://api.subquery.network/sq/agoric-labs/mainnet__YWdvc';