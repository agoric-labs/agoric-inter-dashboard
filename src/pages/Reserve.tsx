import useSWR from 'swr';
import axios, { AxiosError, AxiosResponse } from 'axios';
import { PageHeader } from '@/components/PageHeader';
import { PageContent } from '@/components/PageContent';
import { ReserveSummary } from '@/widgets/ReserveSummary';
import { ReserveShortfall } from '@/widgets/ReserveShortfall';
import { ReserveHistory } from '@/widgets/ReserveHistory';
import { subQueryFetcher, subQueryGraphFetcher } from '@/utils';
import { RESERVE_DASHBOARD_QUERY, RESERVE_GRAPH_TOKENS_QUERY, RESERVE_DAILY_METRICS_QUERY } from '@/queries';
import { ReserveCosmosSummary } from '@/widgets/ReserveCosmosSummary';
import {
  GET_ACCOUNT_BALANCE_URL,
  GET_MODULE_ACCOUNTS_URL,
  VBANK_RESERVE_ACCOUNT,
  UIST_DENOMINATION,
  GRAPH_DAYS,
} from '@/constants';

type GraphData = { key: number; x: string; [key: string]: any };

type OraclePriceNode = {
  typeInAmount: number;
  typeOutAmount: number;
  typeInName: string;
  typeOutName: string;
  id: string;
};
type AllocationsNode = { id: string; token: string; value: number };
type ReserveMetricsNode = {
  allocations: { nodes: Array<AllocationsNode> };
  shortfallBalance: number;
};
type ReserveDashboardResponse = {
  oraclePrices: { nodes: Array<OraclePriceNode> };
  reserveMetrics: { nodes: Array<ReserveMetricsNode> };
};
export type ReserveDashboardData = Array<{
  shortfallBalance: number;
  allocations: { [key: string]: AllocationsNode & OraclePriceNode };
}>;

type ReserveManagerMetricsResponse = {
  reserveMetrics: { nodes: Array<ReserveMetricsNode> };
};

export const Reserve = () => {
  const { data, isLoading } = useSWR<AxiosResponse, AxiosError>(RESERVE_DASHBOARD_QUERY, subQueryFetcher);

  const response: ReserveDashboardResponse = data?.data?.data;
  const oraclePrices: { [key: string]: OraclePriceNode } = response?.oraclePrices?.nodes?.reduce(
    (agg, node) => ({ ...agg, [node.typeInName]: node }),
    {},
  );
  const reserveDashboardQueryData: ReserveDashboardData = response?.reserveMetrics?.nodes?.map((vaultNode) => ({
    ...vaultNode,
    allocations: vaultNode.allocations?.nodes?.reduce(
      (agg, allocation) => ({
        ...agg,
        [allocation.token]: { ...allocation, ...oraclePrices[allocation.token] },
      }),
      {},
    ),
  }));

  const startDate = new Date();
  startDate.setDate(startDate.getDate() - GRAPH_DAYS);
  const startDateFormatDate = startDate.toISOString().slice(0, 10);
  const startDateKey = Number(startDateFormatDate.replaceAll('-', ''));

  //  Queries for graph
  const { data: tokenNamesData } = useSWR<AxiosResponse, AxiosError>(RESERVE_GRAPH_TOKENS_QUERY, subQueryGraphFetcher);
  const tokenNamesResponse: ReserveManagerMetricsResponse = tokenNamesData?.data.data;
  const tokenNames =
    tokenNamesResponse?.reserveMetrics?.nodes
      .map((node) => node.allocations?.nodes?.map((allocation) => allocation.token))
      .flat()
      .toSorted() || [];
  const { data: dailyMetricsData, isLoading: graphDataIsLoading } = useSWR<AxiosResponse, AxiosError>(
    RESERVE_DAILY_METRICS_QUERY(tokenNames, startDateKey),
    subQueryGraphFetcher,
  );
  const dailyMetricsResponse = dailyMetricsData?.data.data;

  const range: number[] = [...Object(Array(90)).keys()].reverse();

  const today = new Date();
  const graphDataMap: { [key: string]: GraphData } = range.reduce((agg, dateNum: number) => {
    const date = new Date(today);
    date.setDate(date.getDate() - dateNum);
    const formatDate = date.toISOString().slice(0, 10);
    const dateKey = Number(formatDate.replaceAll('-', ''));

    return { ...agg, [dateKey]: { key: dateKey, x: formatDate } };
  }, {});

  tokenNames.forEach((tokenName) => {
    const dailyOracles = dailyMetricsResponse?.[`${tokenName}_oracle`]?.nodes.reduce(
      (agg: object, dailyOracleData: { dateKey: string }) => ({ ...agg, [dailyOracleData.dateKey]: dailyOracleData }),
      {},
    );

    let lastTokenMetric = dailyMetricsResponse?.[`${tokenName}_last`]?.nodes[0];

    const dailyMetrics = dailyMetricsResponse?.[tokenName]?.nodes.reduce(
      (agg: object, metricsData: { dateKey: string }) => ({ ...agg, [metricsData.dateKey]: metricsData }),
      {},
    );

    Object.keys(graphDataMap)
      .toSorted()
      .forEach((dateKey) => {
        const oracle = (dailyOracles && dailyOracles[dateKey]) || { typeOutAmountLast: 1, typeInAmountLast: 1 };
        const tokenMetrics = (dailyMetrics && dailyMetrics[dateKey]) || lastTokenMetric;
        graphDataMap[dateKey][tokenName] =
          ((tokenMetrics?.valueLast || 0) / 1_000_000) * (oracle.typeOutAmountLast / oracle.typeInAmountLast);

        lastTokenMetric = tokenMetrics;

      });
  });

  const sortedGraphDataList = Object.values(graphDataMap) as GraphData[];
  sortedGraphDataList.sort((a, b) => a.key - b.key);
  let prevValue: GraphData = sortedGraphDataList[0];
  const graphDataList: Array<GraphData> = sortedGraphDataList
    .reduce((aggArray: Array<GraphData>, graphData: GraphData) => {
      // filling in missing days
      const prevDay = new Date(prevValue.x);
      const nextDay = new Date(graphData.x);
      const timeDiff = nextDay.getTime() - prevDay.getTime();
      const diffDays = Math.ceil(timeDiff / (1000 * 60 * 60 * 24));
      const missingDays =
        diffDays > 1
          ? Array.from(Array(diffDays - 1).keys()).map((idx) => {
              const newDate = new Date(prevDay);
              newDate.setDate(prevDay.getDate() + 1 + idx);
              const newDateString = newDate.toISOString().slice(0, 10);
              const dateKey = Number(newDateString.replaceAll('-', ''));
              return { ...prevValue, x: newDateString, key: dateKey };
            })
          : [];

      const newAggArray = [...aggArray, ...missingDays, { ...prevValue, ...graphData }];
      prevValue = { ...prevValue, ...graphData };
      return newAggArray;
    }, [])
    .slice(-1 * GRAPH_DAYS);

  // Cosmos reserve balance
  const { data: moduleAccounts } = useSWR<AxiosResponse, AxiosError>(GET_MODULE_ACCOUNTS_URL, axios.get);
  const reserveAccount = moduleAccounts?.data.accounts.find(
    (account: { name: string }) => account.name === VBANK_RESERVE_ACCOUNT,
  );
  const reserveAddress = reserveAccount?.base_account.address;

  const { data: reserveBalance, isLoading: reserveBalanceLoading } = useSWR<AxiosResponse, AxiosError>(
    GET_ACCOUNT_BALANCE_URL(reserveAddress),
    axios.get,
  );
  const istReserve: { amount: number } = reserveBalance?.data.balances.find(
    (balance: { denom: string }) => balance.denom === UIST_DENOMINATION,
  );
  const istReserveBalance = (istReserve?.amount || 0) / 1_000_000;

  return (
    <>
      <PageHeader title="Reserve Assets" />
      <PageContent>
        <div className="grid grid-cols-1 md:grid-cols-3 gap-4">
          <ReserveSummary data={reserveDashboardQueryData} isLoading={isLoading} />
          <ReserveCosmosSummary data={istReserveBalance} isLoading={reserveBalanceLoading} />
          <ReserveShortfall data={reserveDashboardQueryData} isLoading={isLoading} />
        </div>
        <ReserveHistory data={graphDataList} tokenNames={tokenNames} isLoading={graphDataIsLoading} />
      </PageContent>
    </>
  );
};
