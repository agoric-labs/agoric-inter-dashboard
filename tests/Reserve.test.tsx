import React from 'react';
import useSWR from 'swr';
import { render } from '@testing-library/react';
import { Reserve } from '@/pages/Reserve';
import { RESERVE_DASHBOARD_QUERY, RESERVE_GRAPH_TOKENS_QUERY } from '@/queries';
import {
  graphDataMock,
  dashboardDataMock,
  graphNodesDataMock,
  moduleAccountsMock,
  reserveBalanceMock,
} from './__mocks__/Reserve';
import { GET_ACCOUNT_BALANCE_URL, GET_MODULE_ACCOUNTS_URL } from '@/constants';

jest.mock('swr', () => ({
  default: jest.fn(),
}));


global.ResizeObserver = jest.fn().mockImplementation(() => ({
  observe: jest.fn(),
  unobserve: jest.fn(),
  disconnect: jest.fn(),
}));
jest.useFakeTimers().setSystemTime(new Date('2024-06-07'));

jest.mock('recharts', () => {
  const OriginalModule = jest.requireActual('recharts');
  return {
    ...OriginalModule,
    ResponsiveContainer: ({ children }: { children: typeof React.Children }) => (
      <OriginalModule.ResponsiveContainer width={800} height={800}>
        {children}
      </OriginalModule.ResponsiveContainer>
    ),
  };
});
describe('Reserve Dashboard Snapshot tests', () => {
  it('should match snapshot when data is loaded', async () => {
    (useSWR as any).mockImplementation((query: string) => {
      const reserveAddress =
        moduleAccountsMock.data.accounts.find((account) => account.name === 'vbank/reserve')?.base_account.address ||
        '';
      const mockData =
        {
          [RESERVE_DASHBOARD_QUERY]: dashboardDataMock,
          [RESERVE_GRAPH_TOKENS_QUERY]: graphNodesDataMock,
          [GET_MODULE_ACCOUNTS_URL]: moduleAccountsMock,
          [GET_ACCOUNT_BALANCE_URL(reserveAddress)]: reserveBalanceMock,
        }[query] || graphDataMock;
      return { data: mockData, error: null, isLoading: false };
    });

    const { container } = render(<Reserve />);
    expect(container).toMatchSnapshot();
  });

  it('should match snapshot when data is loading', async () => {
    (useSWR as any).mockImplementation(() => ({ data: undefined, error: null, isLoading: true }));

    const { container } = render(<Reserve />);
    expect(container).toMatchSnapshot();
  });
});
