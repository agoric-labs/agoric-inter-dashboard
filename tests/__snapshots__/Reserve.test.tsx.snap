// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Reserve Dashboard Snapshot tests should match snapshot when data is loaded 1`] = `
<div>
  <h1
    class="flex items-center space-x-4 lg:space-x-6 border-b p-8 text-3xl font-bold tracking-tight"
  >
    Reserve Assets
  </h1>
  <div
    class="p-8"
  >
    <div
      class="grid grid-cols-1 md:grid-cols-3 gap-4"
    >
      <div
        class="rounded-lg border bg-card text-card-foreground shadow-sm"
      >
        <div
          class="p-6 flex flex-row items-center justify-between space-y-0 pb-2"
        >
          <h3
            class="tracking-tight text-sm font-medium"
          >
            Total Reserve Assets
          </h3>
        </div>
        <div
          class="p-6 pt-0"
        >
          <div
            class="text-2xl font-bold"
          >
            $99.07K
          </div>
        </div>
      </div>
      <div
        class="rounded-lg border bg-card text-card-foreground shadow-sm"
      >
        <div
          class="p-6 flex flex-row items-center justify-between space-y-0 pb-2"
        >
          <h3
            class="tracking-tight text-sm font-medium"
          >
            Cosmos Reserve
          </h3>
        </div>
        <div
          class="p-6 pt-0"
        >
          <div
            class="text-2xl font-bold"
          >
            $33.35K
          </div>
        </div>
      </div>
      <div
        class="rounded-lg border bg-card text-card-foreground shadow-sm"
      >
        <div
          class="p-6 flex flex-row items-center justify-between space-y-0 pb-2"
        >
          <h3
            class="tracking-tight text-sm font-medium"
          >
            Reserve Shortfall
          </h3>
        </div>
        <div
          class="p-6 pt-0"
        >
          <div
            class="text-2xl font-bold"
          >
            $7.84K
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Reserve Dashboard Snapshot tests should match snapshot when data is loading 1`] = `
<div>
  <h1
    class="flex items-center space-x-4 lg:space-x-6 border-b p-8 text-3xl font-bold tracking-tight"
  >
    Reserve Assets
  </h1>
  <div
    class="p-8"
  >
    <div
      class="grid grid-cols-1 md:grid-cols-3 gap-4"
    >
      <div
        class="rounded-lg border bg-card text-card-foreground shadow-sm"
      >
        <div
          class="p-6 flex flex-row items-center justify-between space-y-0 pb-2"
        >
          <h3
            class="tracking-tight text-sm font-medium"
          >
            Total Reserve Assets
          </h3>
        </div>
        <div
          class="p-6 pt-0"
        >
          <div
            class="text-2xl font-bold"
          >
            <div
              class="animate-pulse bg-muted w-[100px] h-[32px] rounded-full"
            />
          </div>
        </div>
      </div>
      <div
        class="rounded-lg border bg-card text-card-foreground shadow-sm"
      >
        <div
          class="p-6 flex flex-row items-center justify-between space-y-0 pb-2"
        >
          <h3
            class="tracking-tight text-sm font-medium"
          >
            Cosmos Reserve
          </h3>
        </div>
        <div
          class="p-6 pt-0"
        >
          <div
            class="text-2xl font-bold"
          >
            <div
              class="animate-pulse bg-muted w-[120px] h-[32px] rounded-full"
            />
          </div>
        </div>
      </div>
      <div
        class="rounded-lg border bg-card text-card-foreground shadow-sm"
      >
        <div
          class="p-6 flex flex-row items-center justify-between space-y-0 pb-2"
        >
          <h3
            class="tracking-tight text-sm font-medium"
          >
            Reserve Shortfall
          </h3>
        </div>
        <div
          class="p-6 pt-0"
        >
          <div
            class="text-2xl font-bold"
          >
            <div
              class="animate-pulse bg-muted w-[100px] h-[32px] rounded-full"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
